### ISSUE TITLE
[Feature request] Export nodes with all their files inside

### ISSUE THREAD

**Issue by zDEFz** [zadam/trilium#1367]
Trillium Notes 
Version: 0.45.2
---

As title

**Comment by zadam** [comment]
If you export as ZIP, then the whole subtree is exported - what do you miss?

**Comment by zDEFz** [comment]
The files inside. All of them are missing. I also want my imported Music, Video etc. files.

**Comment by zadam** [comment]
Perhaps I'm missing something, but the ZIP export should contain all notes - text notes, file notes, image notes.

**Comment by zDEFz** [comment]
I can only assume what happened here. It does no longer happen.
I have around 1GB in the root node, earlier this day when I have exported, the Archive was 472KB in size.
When I have re-opened Trillium Notes, it loaded from a different directory than it was supposed to.

This happened multiple times, but have stopped since I nailed the Path. 
https://github.com/zadam/trilium/wiki/Setting-Database-Path

I can only assume that before it has exported an empty node.
I know macOS is not officially supported, but there seems to be a hiccup, not loading from the proper path. My entry fixed it.

**Comment by zadam** [comment]
Hmm, not sure what the problem is.

Setting data directory is described in here: https://github.com/zadam/trilium/wiki/Data-directory#changing-the-location-of-data-directory

I've merged your Mac specific info into that page and deleted the original.

**Comment by zDEFz** [comment]
zadam, thanks. 
Seems that setting the PATH manually to the defaults fixed that problem.

Would it be possible to programmatically save the .plist file in the directory I proposed?
Just because without that, Trillium seemed to (to the wrong folder) for no reason and then the database appeared empty.
The way it's defined, it automatically loads after login of the user.
With that, you would also have an option for the user to manually pick the folder, where everything is stored, if you wanted.

In the version I started with, Trillium saved to `/Users/def/Library/Application Support/trilium-data`
But for a weird reason, Trillium would save to `/Users/def/trilium-data` when my "Custom path" isn't set.
That path loads by default when the Trilium Path isn't specified or empty, while the default should be `/Users/def/Library/Application Support/trilium-data` Replace def with <user>



Also, thanks for merging my commit.

**Comment by zadam** [comment]
> I assume it loaded an empty db for export, possible?

No, not really. If it didn't find a DB, then during startup it would show wizard to create new document/database. It's not possible that the app would show one database but export a different one.

> In the version I started with, Trillium saved to `/Users/def/Library/Application Support/trilium-data`  
> But for a weird reason, Trillium would save to `/Users/def/trilium-data` when my "Custom path" isn't set.

That is strange, because `/Users/def/Library/Application Support/trilium-data` should be the default on Mac. The only possibility I can think of is that somehow OS is not detected as Mac or that `/Users/def/Library/Application Support` does not exist, in those cases fallback is home directory.

**Comment by zDEFz** [comment]
I see. If the PATH is not found, it would be clever to ask the user to specify it, or give another message.

**Comment by zDEFz** [comment]
@zadam is there any energy going into not loading the fall-back path? I must wait around 3 minutes after my mac has been logged into, to automatically load the script I posted in the wiki. Otherwise, the trilium database appears empty.

**Comment by zadam** [comment]
No, I'm not able to reproduce this problem so can't fix it.

What about creating a symlink or moving the directory to the expected place?

**Comment by zDEFz** [comment]
Tell me what I should do, I don't understand.

**Comment by zadam** [comment]
Trilium is looking for some reason in `/Users/def/trilium-data` while it should look at `/Users/def/Library/Application Support/trilium-data`

You can either simply move the directory from the latter to former or you can create a symbolic link from latter to former ...

**Comment by zDEFz** [comment]
I'm more asking what I should do for you so it is fixed in the code, since you said, you can't test it yourself.

**Comment by zadam** [comment]
Frankly, I'm not convinced yet that this is a bug as opposed to something going wrong in your system. This code/logic is pretty old, probably more than a year and it seems to work fine for all the other users ...

**Comment by zDEFz** [comment]
I think a simple fix would be to load always from the same path, not using a fallback. Let the user specify the path manually in the config. That would do. Also, u no longer need to specify it in the .plist file.

**Comment by zadam** [comment]
Not using a fallback would break existing users who depend on that fallback so that's not an option.

> Let the user specify the path manually in the config.

First you need to find that config file ... For that reason the path is in environment variable.

**Comment by zDEFz** [comment]
Yea but the environment variable needs to get loaded each time.

Since a recent version of macOS you have .plist files...

And things like `export JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home` or so are from Sierra times. 

That no longer works on the newer macOS versions.

It's like I said changed to ... `launchctl setenv TRILIUM_DATA_DIR <yourpath>` 
and the path should be in doublequotes.

**Comment by zadam** [comment]
> Can't you make a new case in your switch statement which I assume you have, and instead of saying `export`, you say `launchctl setenv...`

There isn't anything like that in trilium code - that's what needs to happen outside of the application.

I don't have much knowledge regarding this plist and frankly I'm a bit baffled why Apple decided to make it harder (and that you need to do it after every restart, why?) But because of these Apple specific things I explicitly don't support Mac platform. I do provide the binaries, but that's it.

**Comment by zDEFz** [comment]
@zadam I tried to run your software in Docker, and I gave it 32 GB of RAM + 4 cpu cores. Yet it was much slower than the application. That's my reason to run it in the Application...

I don't know how other software manufacturers are doing this thing with the path (where it access to), but maybe someone else has an Idea.
The big problem is just, that for me, if I don't touch it, my database appears empty on startup. I don't have an localized OS, at all.

I would suggest that I open a new issue and tag it with [Help needed]
